### PR TITLE
shell: fix log output when CONFIG_LOG_IMMEDIATE=y

### PR DESCRIPTION
**Pull request by jimparis**
With CONFIG_LOG_IMMEDIATE, log output is printed immediately.  If a shell command is in progress, there's no prompt to erase, nor should we print a new prompt after the log message is output.

Before this patch, a simple shell command like:
```
static int cmd_log_erase(const struct shell *shell, size_t argc, char **argv)
{
        LOG_INF("hello world");
        return 0;
}
```
would output something like:

```
uart:~$ log erase
[00:00:02.623,718] <inf> cmd_log: hello world
uart:~$ loguart:~$
```

### CODE REVIEW

File: subsys/shell/shell_log_backend.c [hunk 1]
```diff
@@ -245,7 +249,9 @@ static void put_sync_hexdump(const struct log_backend *const backend,
 	}
 
 	key = irq_lock();
-	shell_cmd_line_erase(shell);
+	if (!flag_cmd_ctx_get(shell)) {
+		shell_cmd_line_erase(shell);
+	}
 	log_output_hexdump(shell->log_backend->log_output, src_level, timestamp,
 			   metadata, data, length, flags);
 	irq_unlock(key);
```
Comment (nordic-krch) [inline subsys/shell/shell_log_backend.c]: Can you also add here
```
if (!flag_cmd_ctx_get(shell)) {
		shell_print_prompt_and_cmd(shell);
	}
```
and remove:
```/* Even though log message is handled notify shell thread which
	 * will print command buffer after the log message.
	 */
	if (IS_ENABLED(CONFIG_MULTITHREADING)) {
		signal = &shell->ctx->signals[SHELL_SIGNAL_LOG_MSG];
		k_poll_signal_raise(signal, 0);
	}
```
i'm not sure how it happened that this functions are different in the way they handle things. They should be the same, except for log_output function.

**Comment by nordic-krch** [comment]
Thanks for this patch. I've left one comment.

File: subsys/shell/shell_log_backend.c (hunk 1 above)
Comment (jimparis) [reply subsys/shell/shell_log_backend.c]: Updated

**Review by nordic-krch — APPROVED** [review]
thanks. looks good (with minor thing reported by CI).

**Comment by jimparis** [comment]
> thanks. looks good (with minor thing reported by CI).

fixed

**Review by tyrantlizard77 — APPROVED** [review]

**Comment by ioannisg** [comment]
Does this need to get back-ported to 1.14? @jimparis, @nordic-krch ?

**Comment by jakub-uC** [comment]
@ioannisg : It is not a severe problem. In my opinion it is not a must be for back porting.

**Review by jakub-uC — APPROVED** [review]

**Pull request merged**